include/podio/detail/LinkCollectionIterator.h

```diff
@@ -0,0 +1,47 @@
+#ifndef PODIO_DETAIL_LINKCOLLECTIONITERATOR_H
+#define PODIO_DETAIL_LINKCOLLECTIONITERATOR_H
+
+#include "podio/detail/LinkFwd.h"
+#include "podio/utilities/MaybeSharedPtr.h"
+
+namespace podio {
+template <typename FromT, typename ToT, bool Mutable>
+class LinkCollectionIteratorT {
+  using LinkType = LinkT<FromT, ToT, Mutable>;
+  using LinkObjT = LinkObj<FromT, ToT>;
+
+public:
+  LinkCollectionIteratorT(size_t index, const LinkObjPointerContainer<FromT, ToT>* coll) :
+      m_index(index), m_object(podio::utils::MaybeSharedPtr<LinkObjT>{nullptr}), m_collection(coll) {
+  }
+
+  LinkCollectionIteratorT(const LinkCollectionIteratorT&) = delete;
+  LinkCollectionIteratorT& operator=(const LinkCollectionIteratorT&) = delete;
+
+  bool operator!=(const LinkCollectionIteratorT& other) const {
+    return m_index != other.m_index; // TODO: may not be complete
+  }
+
+  LinkType operator*() {
+    m_object.m_obj = podio::utils::MaybeSharedPtr<LinkObjT>((*m_collection)[m_index]);
+    return m_object;
+  }
+
+  LinkType* operator->() {
```

Comment:
what's the const qualifier here?

Reply:
They cannot be easily marked `const` because we switch the object that we return during the access (and there is no other place where we could do this). I don't think it's required by anything that these access operators are `const`.

Reply:
OK. Fine then.